zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -79,30 +75,33 @@ pub(crate) mod address_high {
 /// the RLC value for `memory_offset` need not match the bytes.
 #[derive(Clone, Debug)]
 pub(crate) struct MemoryAddressGadget<F> {
-    memory_offset: Word32Cell<F>,
-    memory_length: Word32Cell<F>,
-    memory_length_is_zero: IsZeroWordGadget<F, Word32Cell<F>>,
+    memory_offset_bytes: MemoryAddress<F>,
+    memory_offset_word: WordCell<F>,
+    memory_length: MemoryAddress<F>,
+    memory_length_is_zero: IsZeroGadget<F>,
 }
 
 impl<F: Field> MemoryAddressGadget<F> {
-    #[deprecated(note = "construct_new is favored")]
     pub(crate) fn construct(
-        _cb: &mut EVMConstraintBuilder<F>,
-        _memory_offset: Cell<F>,
-        _memory_length: MemoryAddress<F>,
-    ) -> Self {
-        todo!()
-    }
-
-    pub(crate) fn construct_new(
         cb: &mut EVMConstraintBuilder<F>,
-        memory_offset: Word32Cell<F>,
-        memory_length: Word32Cell<F>,
+        memory_offset_word: WordCell<F>,
+        memory_length: MemoryAddress<F>,
```

Comment:
I wonder if `memory_length` should be `Word32` for these two reasons:
- Usually these arguments come from the stack, and the user may put anything there.  Having a length too big will case an error but we're actually using these gadgets in Error states, so we need to be able to read the stack value and prove it's too big.  Perhaps we can get away with it:  The ExecStep could allocate Word32, and from it build a MemoryAddress (by taking the less significant 5 bytes).

After thinking a bit more about this: maybe it's OK to use `MemoryAddress` here:
- Allocating MemoryAddress for length and reading it from stack in an opcode ExecStep should be OK, because we only prove that when the step succeeds, so it should fit.
- For the ExecStates that do OOG due to length being more than 5 bytes, we can't use `MemoryAddress` for the stack read of length.